### ISSUE TITLE
Indentation formatting

### ISSUE THREAD

**Issue by panmareksadowski** [jcrist/msgspec#219]
Hi,

Is indentation formatting supported by the msgspec json encoder as standard python json supports, or is support planned to be added?

**Comment by jcrist** [comment]
Indentation formatting is not supported, and I wasn't planning on adding it.

`msgspec` is designed to be as performant as possible; one of the first things people do when trying to improve JSON performance is strip whitespace - do you have a use case where you need performant JSON encoding with indentation?

If not, you can handle this by mixing msgspec with the standard library:

```python
import msgspec
import json


def pretty_json(obj):
    """Pretty-print JSON, supporting all types that msgspec supports"""

    # This will be slower than a single `msgspec.json.encode`, but if you're
    # pretty-printing JSON then you usually don't care as much about performance

    def default(obj):
        """Roundtrip stdlib unsupported types through msgspec"""
        return msgspec.json.decode(msgspec.json.encode(obj))

    return json.dumps(
        obj,
        indent=2,
        check_circular=False,  # not needed, can't have a circular structure
        ensure_ascii=False,    # no need to escape unicode characters
        default=default,
    )


class Point(msgspec.Struct):
    x: int
    y: int


# Some msgspec-specific message types, unsupported by the stdlib JSON
msg = [Point(1, 2), Point(3, 4), Point(5, 6)]


print(pretty_json(msg))
```

Output

```
[
  {
    "x": 1,
    "y": 2
  },
  {
    "x": 3,
    "y": 4
  },
  {
    "x": 5,
    "y": 6
  }
]
```

**Comment by panmareksadowski** [comment]
In my use case, decoding performance is crucial and it was the reason for choosing msgspec as I did not find anything comparable in performance (really awesome work).
On the other hand, encoding performance is less important in my use case, but I still would like to work with msgspec.Struct objects, and sometimes have the ability to format json to be more human readable. I have tried just `json.dumps(json.loads(msgspec.json.encode(obj)), indent=2)`, and it works, but it is overkill for me to encode, decode and encode again. Even if performance is less important it increased encoding time from 4s to 80s and it is not acceptable in my case. I have tried what you proposed, but it does just more granularity the same: encode, decode, and encode again. The performance is the same as encoding, decoding, and encoding again once at the top level.

**Comment by jcrist** [comment]
> Even if performance is less important it increased encoding time from 4s to 80s

Woah, even `4s` is a long time for a `msgspec.json.encode` call (and 80s to roundtrip is huge). This must be a very large JSON document - how large is it?

```python
len(msgspec.json.encode(your_message))
```

>  and sometimes have the ability to format json to be more human readable.

I can sympathize with this use case. There *may* be a way to support optional indentation with negligible slowdown in the encoder. I'd need to iterate on this a bit to ensure there are no performance regressions.

However, we could also offer an option to prettify a JSON message without decoding it. This would be a function with a signature like:

```python
def format_json(msg: bytes, indent: int = 2) -> bytes:
    """Format an existing JSON message to be more human readable"""
    ...
```

The changes needed to write such a function wouldn't affect performance of msgspec's existing JSON encoder, and I _think_ would be pretty straightforward. This wouldn't offer as high performance as implementing indentation directly in `msgspec.json.encode`, but the overhead shouldn't be that high. The main cost here is a doubling+ of memory usage, since you have both the original (unformatted) and new (formatted) messages in memory at the same time.

Would a function like that satisfy your use case?

**Comment by panmareksadowski** [comment]
> Woah, even `4s` is a long time for a `msgspec.json.encode` call (and 80s to roundtrip is huge). This must be a very large JSON document - how large is it?
> 
> ```python
> len(msgspec.json.encode(your_message))
> ```
Yes, it is quite big, around 800MB.  `len(msgspec.json.encode(your_message)) == 833654479`

> Would a function like that satisfy your use case?
Completely. Memory usage is not a problem in my case.

**Comment by panmareksadowski** [comment]
I tried it and it works perfectly for me. The time of formatting with indentation takes additional around the same time as encoding, but it is acceptable and incomparably better than the roundtrip of additional encoding-decoding.

P.S.
> Woah, even 4s is a long time for a msgspec.json.encode call (and 80s to roundtrip is huge). This must be a very large JSON document - how large is it?

I could mislead you a little with that time as I counted the time together with writing to file, but we have fast drives. The only msgspec encoding time was half of that, so ~2s . Anyway, it didn't change the whole picture (omitting the time of writing to file, json encoding-decoding roundtrip obviously gives even more relative overhead).
